src/metabase/public_settings/premium_features.clj

```diff
@@ -240,3 +241,169 @@
+  `classloader/require` and `ns-resolve`."
+  (memoize
+   (fn [ee-ns fn-name]
+     (u/ignore-exceptions
+      (classloader/require (symbol ee-ns))
+      (ns-resolve (symbol ee-ns) (symbol fn-name))))))
+
+(defmacro defenterprise-oss
+  "Impl macro for `defenterprise` when used in an OSS namespace. Don't use this directly."
+  [{:keys [fn-name ee-ns docstr args body result-schema]}]
+  (register-mapping! fn-name ee-ns args body)
+  `(if-let [ee-fn# (resolve-ee ~(str ee-ns) ~(str fn-name))]
+     (apply ee-fn# ~args)
+     (let [result# (do ~@body)]
+       (if ~result-schema
+         (s/validate ~result-schema result#)
+         result#))))
+
+(defmacro defenterprise
+  "Defines a function that has separate implementations between the Metabase Community Edition (CE, aka OSS) and
+  Enterprise Edition (EE).
+
+  When used in a CE namespace, defines a function that should have a corresponding implementation in an EE namespace
+  (using the same macro). The EE implementation will be used preferentially to the CE implementation if it is available.
+  The first argument after the function name should be a symbol of the namespace containing the EE implementation. The
+  corresponding EE function must have the same name as the CE function.
+
+  When used in an EE namespace, the namespace of the corresponding CE implementation does not need to be included --
+  it will be inferred automatically, as long as a corresponding [[defenterprise]] call exists in an CE namespace.
+
+  Two additional options can be defined, when using this macro in an EE namespace. These options should be defined
+  immediately before the args list of the function:
+
+  ###### `:feature`
+
+  A keyword representing a premium feature which must be present for the EE implementation to be used. Use `:any` to
+  require a valid premium token with at least one feature, but no specific feature. Use `:none` to always run the
+  EE implementation if available, regardless of token. (Default: `:any`)
+
+  ###### `:fallback`
+
+  A keyword or function representing the fallback mechanism which should be used if the instance does not have the
+  premium feature defined by the :feature option. If a function is provided, it will be called with the same args
+  as the EE function. Valid keyword options are `:error`, which causes an exception to be thrown, or `:oss`, which
+  causes the CE implementation of the function to be called. (Default: `:oss`)"
+  [fn-name & defenterprise-args]
+  {:pre [(symbol? fn-name)]}
+  (let [{:keys [docstr args arg->schema result-schema]
+         :as defenterprise-args} (-> (parse-defenterprise-args defenterprise-args)
+                                     (assoc :fn-name fn-name))]
+    (when-not (:docstr defenterprise-args)
+      (log/warn (u/format-color 'red "Warning: enterprise function %s/%s does not have a docstring. Go add one."
+                  (ns-name *ns*) fn-name)))
```

Comment:
You might only want to do this if we're in dev mode (check `config/is-dev?`) so we don't have people bothering Flamber with "why is there this warning in the logs" in prod

Reply:
I copied this message over from `defendpoint`...should I make this change there as well? Doesn't seem like that one is gated by dev mode either.

Reply:
Yeah `defendpoint` should definitely only be logging it in dev mode as well

Reply:
Another options is just throwing an error if there's no docstring. That's what we do with `defsetting`. A lot harder to ignore than a logged warning only in dev